xarray/core/coordinates.py

```diff
@@ -314,6 +315,71 @@ def __unicode__(self):
         return formatting.indexes_repr(self)
 
 
+def normalize_indexes(indexes, coords, sizes):
+    """Normalize indexes for Dataset/DataArray.
+
+    - Validates that all indexes are pd.Index instances (or at least satisfy
+      the Index API we need for xarray).
+    - Combines indexes along the same dimension into a MultiIndex.
+    - Creates default indexes for variables whose name matches their sole
+      dimension.
+
+    Parameters
+    ----------
+    indexes : Optional[Dict[Any, pandas.Index]]
+        Explicitly supplied indexes, if any.
+    coords : Mapping[Any, xarray.Variable]
+        Coordinate variables from which to draw default indexes.
+    dim_sizes : Mapping[Any, int]
+        Integer sizes for each Dataset/DataArray dimension.
+
+    Returns
+    -------
+    Mapping[Any, pandas.Index] mapping indexing keys (levels/dimension names)
+    to indexes used for indexing along that dimension.
+    """
+    indexes = {} if indexes is None else dict(indexes)
+
+    # default indexes
+    for key in sizes:
```

Comment:
Do we want to allow the sizes to be optional, and drawn from the indexes / coords? At the moment they must be supplied in `sizes`

Reply:
I was imagining calling this function *after* the Dataset/DataArray is filled in with fully specified `xarray.Variable` objects, so we would already have determined sizes for all dimensions.

Reply:
Perfect, that makes sense

Reply:
Eventually (and I realize from the other PR this is a sketch atm) then we likely want to check the sizes are consistent if they're also supplied in `indexes` or `coords`